### PR TITLE
Added TypeTranslationRepository

### PR DESCRIPTION
**Pull request by david-buderus**

### CODE REVIEW

**Comment by bastihav** [comment]
Please test the `insertAll` method

File: server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    if (optTranslation.isEmpty()) {
+      return insert(universe, new ItemTypeTranslation(null, type, broaderVariants));
+    }
+
+    ItemTypeTranslation typeTranslation = optTranslation.get();
+    Set<ItemType> variants = new HashSet<>(broaderVariants);
+    variants.addAll(typeTranslation.getBroaderVariants());
+
+    if (variants.size() == typeTranslation.getBroaderVariants().size()) {
+      // No new types where added
```
Comment (bastihav) [inline server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java]:
```suggestion
      // No new types were added
```

File: server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+   */
+  public ItemTypeTranslation addTypeTranslation(String universe, ItemType type,
+      Set<ItemType> broaderVariants) {
+    Optional<ItemTypeTranslation> optTranslation = get(universe, type);
+
+    if (optTranslation.isEmpty()) {
+      return insert(universe, new ItemTypeTranslation(null, type, broaderVariants));
+    }
+
+    ItemTypeTranslation typeTranslation = optTranslation.get();
```
Comment (bastihav) [inline server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java]: rename to `existingTypeTranslation`

File: server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      return typeTranslation;
+    }
+
+    return update(universe, typeTranslation.getId(),
+        new ItemTypeTranslation(typeTranslation.getId(), typeTranslation.getType(),
+            variants));
+  }
+
+  /**
+   * Returns all broader variants of the given name.
```
Comment (bastihav) [inline server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java]:
```suggestion
   * Returns all broader variants of the given type.
```

File: server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+      if (alreadySeen.contains(currentType)) {
+        continue;
+      }
+      result.add(currentType);
+      alreadySeen.add(currentType);
+
+      get(universe, currentType).ifPresent(
+          typeTranslation -> queue.addAll(typeTranslation.getBroaderVariants()));
+    }
```
Comment (bastihav) [inline server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java:95]: use a do-while to get rid of lines 81 and 82?

File: server/src/main/java/de/pnp/manager/server/database/RepositoryBase.java [hunk 5]
```diff
@@ -136,4 +145,22 @@ protected MongoTemplate getTemplate(String universe) {
...
     return config.mongoTemplate(universe);
   }
+
+  /**
+   * Possibility to add validations or something similar before the repository tries to persist the
+   * object.
+   */
+  protected void onPrePersistent(String universe, E object) {
+    // no op
+  }
```
Comment (bastihav) [inline server/src/main/java/de/pnp/manager/server/database/RepositoryBase.java]: rename to `onBeforePersist()`

File: server/src/main/java/de/pnp/manager/server/database/RepositoryBase.java [hunk 6]
```diff
@@ -136,4 +145,22 @@ protected MongoTemplate getTemplate(String universe) {
...
+
+  /**
+   * Possibility to add hooks or something similar after the repository has successfully persisted
+   * the object.
+   * <p>
+   * This means the object will always have an {@link DatabaseObject#getId() id}.
+   */
+  protected void onPostPersistent(String universe, E object) {
+    // no op
+  }
```
Comment (bastihav) [inline server/src/main/java/de/pnp/manager/server/database/RepositoryBase.java]: `onAfterPersist`

File: server/src/test/java/de/pnp/manager/server/component/ItemBuilder.java [hunk 7]
```diff
@@ -2,38 +2,67 @@
...
+
+    @Autowired
+    private ItemTypeRepository typeRepository;
+
+    /**
+     * Builder with default values.
+     */
+    public ItemBuilder someItem(String universe) {
+      return new ItemBuilder(universe, typeRepository);
+    }
```
Comment (bastihav) [inline server/src/test/java/de/pnp/manager/server/component/ItemBuilder.java]: ????

File: server/src/test/java/de/pnp/manager/server/component/ItemBuilder.java [hunk 8]
```diff
@@ -2,38 +2,67 @@
...
-  protected int armor;
-  protected double weight;
-
-  public static ItemBuilder someItem() {
-    return new ItemBuilder();
+  /**
+   * Component wrapper for the {@link ItemBuilder}.
+   */
+  @Component
+  public static class ItemBuilderFactory {
```
Comment (bastihav) [inline server/src/test/java/de/pnp/manager/server/component/ItemBuilder.java]: Rename to make it obvious that this is a test utils class.

File: server/src/test/java/de/pnp/manager/server/database/RepositoryTestBase.java [hunk 9]
```diff
@@ -77,6 +89,64 @@ void testUpdate() {
...
+    linkRepository.update(universe, originalLink.getId(), changedLink);
+    assertThat(linkRepository.get(universe, originalLink.getId())).contains(changedLink);
+
+    Optional<E> optional = repository.get(universe, persistedObject.getId());
+    assertThat(optional).isNotEmpty();
+    assertThat(linkGetter.apply((A) optional.get())).isEqualTo(changedLink);
+  }
+
+  /**
+   * A helper method to tests links between to repositories.
```
Comment (bastihav) [inline server/src/test/java/de/pnp/manager/server/database/RepositoryTestBase.java]:
```suggestion
   * A helper method to test links between two repositories.
```

**Review by bastihav — CHANGES_REQUESTED** [review]

File: server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java (hunk 4 above)
Comment (david-buderus) [reply server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java:95]: The queue could be empty at the start of the loop. I would need to add an if with a break.
IMHO is this solution better.

Comment (bastihav) [reply server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java:95]:
```suggestion
    ItemType currentType = type;
    do {
      if (alreadySeen.contains(currentType)) {
          currentType = queue.poll();
          continue;
      }
      result.add(currentType);
      alreadySeen.add(currentType);
      get(universe, currentType).ifPresent(
          typeTranslation -> queue.addAll(typeTranslation.getBroaderVariants()));

      currentType = queue.poll();
    } while (currentType != null);
```
What about this?

Comment (bastihav) [reply server/src/main/java/de/pnp/manager/server/database/ItemTypeTranslationRepository.java:95]: Eh I guess we do not win anything there.

**Comment by bastihav** [comment]
Please merge develop.

**Pull request merged**